File: 1129.py
from collections import deque

class Solution(object):
    def shortestAlternatingPaths(self, n, redEdges, blueEdges):
        """
        :type n: int
        :type redEdges: List[List[int]]
        :type blueEdges: List[List[int]]
        :rtype: List[int]
        """
        red = [[] for _ in range(n)]
        blue = [[] for _ in range(n)]
    
        for u, v in redEdges:
            red[u].append(v)
        for u, v in blueEdges:
            blue[u].append(v)

        res = [-1] * n
        res[0] = 0
        q = deque([(0, 0), (0, 1)])
        dist = [[-1]*2 for _ in range(n)]
        dist[0][0] = dist[0][1] = 0

        while q:
            node, color = q.popleft()
            edges = blue[node] if color else red[node]
        
            for v in edges:
                if dist[v][1-color] == -1:
                    dist[v][1-color] = dist[node][color] + 1
                    if res[v] == -1 or dist[v][1-color] < res[v]:
                        res[v] = dist[v][1-color]
                    q.append((v, 1-color))
    
        return res

"""
n = 3
redEdges = [[0,1]]
blueEdges = [[2,1]]
solution = Solution()
result = solution.shortestAlternatingPaths(n, redEdges, blueEdges)
print(result)
"""

File: 1129.java
import java.util.*;

class Solution {
    private static final int RED = 0, BLUE = 1, INF = 1_000_000_000;

    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {

        List<Integer>[] g = new ArrayList[n];
        List<Integer>[] red = new ArrayList[n], blue = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            red[i] = new ArrayList<>();
            blue[i] = new ArrayList<>();
        }
        for (int[] e : redEdges)  red[e[0]].add(e[1]);
        for (int[] e : blueEdges) blue[e[0]].add(e[1]);

        int[][] dist = new int[n][2];
        for (int[] d : dist) Arrays.fill(d, INF);
        dist[0][RED] = dist[0][BLUE] = 0;

        Deque<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0, RED});
        q.offer(new int[]{0, BLUE});

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int v = cur[0];
            int col = cur[1];
            int d = dist[v][col];

            List<Integer> nextList = (col == RED) ? blue[v] : red[v];
            int nextCol = col ^ 1;

            for (int nxt : nextList) {
                if (dist[nxt][nextCol] == INF) { 
                    dist[nxt][nextCol] = d + 1;
                    q.offer(new int[]{nxt, nextCol});
                }
            }
        }

        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            int best = Math.min(dist[i][RED], dist[i][BLUE]);
            ans[i] = (best == INF) ? -1 : best;
        }
        return ans;
    }
}